checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -4,88 +4,340 @@
- * (via BaseTypeVisitor). It also manually verifies that @Holding and @HoldingOnEntry
- * annotations are properly used on overridden methods. It handles @GuardedBy annotations
- * on method receivers. Finally, it ensures that we avoid doing any lock checking
- * when visiting initializers.
+ * The LockVisitor enforces the special type-checking rules described in the Lock Checker manual chapter.
+ *
+ * @checker_framework.manual #lock-checker Lock Checker
  */
 
 public class LockVisitor extends BaseTypeVisitor<LockAnnotatedTypeFactory> {
-    private final Class<? extends Annotation> checkerGuardedByClass = org.checkerframework.checker.lock.qual.GuardedBy.class;
-    private final Class<? extends Annotation> checkerHoldingClass = org.checkerframework.checker.lock.qual.Holding.class;
-    private final Class<? extends Annotation> checkerHoldingOnEntryClass = org.checkerframework.checker.lock.qual.HoldingOnEntry.class;
-    private final Class<? extends Annotation> checkerLockHeldClass = org.checkerframework.checker.lock.qual.LockHeld.class;
+    private final Class<? extends Annotation> checkerGuardedByClass = GuardedBy.class;
+    private final Class<? extends Annotation> checkerGuardSatisfiedClass = GuardSatisfied.class;
 
     // Note that Javax and JCIP @GuardedBy is used on both methods and objects. For methods they are
     // equivalent to the Checker Framework @Holding annotation.
     private final Class<? extends Annotation> javaxGuardedByClass = javax.annotation.concurrent.GuardedBy.class;
     private final Class<? extends Annotation> jcipGuardedByClass = net.jcip.annotations.GuardedBy.class;
 
+    /** Annotation constants */
+    protected final AnnotationMirror GUARDEDBY, GUARDEDBYINACCESSIBLE, GUARDSATISFIED, GUARDEDBYBOTTOM;
+
     public LockVisitor(BaseTypeChecker checker) {
         super(checker);
 
+        GUARDEDBYINACCESSIBLE = AnnotationUtils.fromClass(elements, GuardedByInaccessible.class);
+        GUARDEDBY = AnnotationUtils.fromClass(elements, GuardedBy.class);
+        GUARDSATISFIED = AnnotationUtils.fromClass(elements, GuardSatisfied.class);
+        GUARDEDBYBOTTOM = AnnotationUtils.fromClass(elements, GuardedByBottom.class);
+
         checkForAnnotatedJdk();
     }
 
     @Override
+    public Void visitVariable(VariableTree node, Void p) { // visit a variable declaration
+        // A user may not annotate a primitive type, a boxed primitive type or a String
+        // with any qualifier from the @GuardedBy hierarchy.
+        TypeMirror nodeType = atypeFactory.getAnnotatedType(node).getUnderlyingType();
+        if ((TypesUtils.isBoxedPrimitive(nodeType) ||
+             TypesUtils.isPrimitive(nodeType) ||
+             TypesUtils.isString(nodeType)) &&
+            (node.toString().contains("GuardSatisfied") ||
+             node.toString().contains("GuardedBy"))){ // HACK!!! TODO: Fix once there is a way to reliably retrieve user-written qualifiers.
```

Comment:
Added a comment. The checking is not special for Strings. Essentially, no primitive or built-in immutable type should have an annotation from the @GuardedBy hierarchy written by the user, so we check for user written annotations here an issue a warning. This is because primitives and built-in immutable types need to be guarded by name, which is a feature that is not implemented yet. In a future release, the text of this warning can be changed to say "please use the @GuardedByName annotation for primitives, boxed primitives and strings." We may or may not provide a mechanism to allow users to specify their own types as being immutable.